src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketBranchDiscoveryTrait.java

```diff
@@ -0,0 +1,120 @@
+package com.atlassian.bitbucket.jenkins.internal.scm;
+
+import com.atlassian.bitbucket.jenkins.internal.client.BitbucketClientFactory;
+import com.atlassian.bitbucket.jenkins.internal.client.BitbucketClientFactoryProvider;
+import com.atlassian.bitbucket.jenkins.internal.client.BitbucketBranchClient;
+import com.atlassian.bitbucket.jenkins.internal.client.BitbucketRepositoryClient;
+import com.atlassian.bitbucket.jenkins.internal.config.BitbucketPluginConfiguration;
+import com.atlassian.bitbucket.jenkins.internal.config.BitbucketServerConfiguration;
+import com.atlassian.bitbucket.jenkins.internal.credentials.JenkinsToBitbucketCredentials;
+import hudson.Extension;
+import jenkins.plugins.git.GitSCMBuilder;
+import jenkins.scm.api.SCMHead;
+import jenkins.scm.api.SCMHeadOrigin;
+import jenkins.scm.api.SCMRevision;
+import jenkins.scm.api.trait.SCMBuilder;
+import jenkins.scm.api.trait.SCMSourceContext;
+import jenkins.scm.api.trait.SCMSourceTraitDescriptor;
+import jenkins.scm.impl.trait.Discovery;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+import javax.inject.Inject;
+import java.util.Optional;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Stream;
+
+public class BitbucketBranchDiscoveryTrait extends BitbucketSCMSourceTrait {
+
+    private static final Logger log = Logger.getLogger(BitbucketBranchDiscoveryTrait.class.getName());
+
+    @DataBoundConstructor
+    public BitbucketBranchDiscoveryTrait() {
+    }
+
+    @Override
+    protected void decorateContext(SCMSourceContext<?, ?> context) {
+        if (context instanceof BitbucketSCMSourceContext) {
+            BitbucketSCMSourceContext bitbucketContext = (BitbucketSCMSourceContext) context;
+
+            DescriptorImpl
+                    descriptor = (DescriptorImpl) getDescriptor();
+            Optional<BitbucketClientFactory> clientFactory =
+                    descriptor.getClientFactory(bitbucketContext);
+
+            if (!clientFactory.isPresent()) {
+                log.log(Level.WARNING, "Server configuration missing, cannot resolve client for Branch discovery");
+                return;
+            }
+
+            BitbucketSCMRepository repository = bitbucketContext.getRepository();
+            BitbucketRepositoryClient repositoryClient = clientFactory.get()
+                    .getProjectClient(repository.getProjectKey())
+                    .getRepositoryClient(repository.getRepositorySlug());
+            BitbucketBranchClient bitbucketBranchClient = repositoryClient.getBranchClient();
```

Comment:
nit: we don't seem to be using `repositoryClient` so we can probably just directly assign `bitbucketBranchClient`
```suggestion
            BitbucketBranchClient bitbucketBranchClient = clientFactory.get()
                    .getProjectClient(repository.getProjectKey())
                    .getRepositoryClient(repository.getRepositorySlug())
                    .getBranchClient();
```